File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/ProjectProvider.java
package org.jetbrains.bsp.bazel.server.sync;

import org.jetbrains.bsp.bazel.server.sync.model.Project;

public class ProjectProvider {
  private Project project;
  private final ProjectResolver projectResolver;
  private final ProjectStorage projectStorage;

  public ProjectProvider(ProjectResolver projectResolver, ProjectStorage projectStorage) {
    this.projectResolver = projectResolver;
    this.projectStorage = projectStorage;
  }

  public synchronized Project refreshAndGet() {
    loadFromBazel();
    return project;
  }

  public synchronized Project get() {
    if (project == null) {
      loadFromDisk();
    }
    if (project == null) {
      loadFromBazel();
    }

    return project;
  }

  private void loadFromBazel() {
    project = projectResolver.resolve();
    storeOnDisk();
  }

  private void loadFromDisk() {
    projectStorage.load().forEach(p -> project = p);
  }

  private void storeOnDisk() {
    projectStorage.store(project);
  }
}

File: server/src/main/kotlin/org/jetbrains/bsp/bazel/server/sync/ProjectProvider.kt
package org.jetbrains.bsp.bazel.server.sync

import org.eclipse.lsp4j.jsonrpc.CancelChecker
import org.jetbrains.bsp.bazel.server.model.Project

class ProjectProvider(private val projectResolver: ProjectResolver) {
    private var project: Project? = null

    @Synchronized
    fun refreshAndGet(cancelChecker: CancelChecker, build: Boolean): Project =
        loadFromBazel(cancelChecker, build = build)

    @Synchronized
    fun get(cancelChecker: CancelChecker): Project = project ?: loadFromBazel(cancelChecker, false)

    private fun loadFromBazel(cancelChecker: CancelChecker, build: Boolean) = projectResolver.resolve(cancelChecker, build = build).also {
        project = it
        System.gc()
    }
}
